Source/WebCore/css/CSSCounterStyle.cpp

```diff
@@ -27,52 +27,211 @@
+    if (!amountOfSymbols) {
+        ASSERT_NOT_REACHED();
+        return { };
+    }
+    if (!value)
+        return symbols().at(0);
+
+    Vector<String> reversed;
+    while (value) {
+        reversed.append(symbols().at(value % amountOfSymbols));
+        value = static_cast<int>(std::floor(value / amountOfSymbols));
+    }
+    StringBuilder result;
+    for (auto iter = reversed.rbegin(); iter != reversed.rend(); ++iter)
+        result.append(*iter);
+    return result.toString();
 }
-String CSSCounterStyle::counterForSystemAdditive(int)
+
+// https://www.w3.org/TR/css-counter-styles-3/#additive-system
+String CSSCounterStyle::counterForSystemAdditive(int value)
 {
-    // FIXME: implement counter for system rdar://103648354
-    return ""_s;
+    auto& aSymbols = additiveSymbols();
+    if (!value) {
+        for (auto& [symb, weight] : aSymbols) {
+            if (!weight)
+                return symb;
+        }
+        return { };
+    }
+
+    StringBuilder result;
+    auto appendToResult = [&](String symb, unsigned frequency) {
+        for (unsigned i = 0; i < frequency; ++i)
+            result.append(symb);
+    };
+
+    for (auto& [symb, weight] : aSymbols) {
+        if (!weight || static_cast<int>(weight) > value)
+            continue;
+        auto repetitions = static_cast<unsigned>(std::floor(value / weight));
+        appendToResult(symb, repetitions);
+        value -=  weight * repetitions;
+        if (!value)
+            return result.toString();
+    }
+
+    return { };
 }
-String CSSCounterStyle::initialRepresentation(int)
+
+String CSSCounterStyle::initialRepresentation(int value)
 {
-    // FIXME: implement counter initial representation rdar://103648354
+    switch (system()) {
+    case CSSCounterStyleDescriptors::System::Cyclic:
+        return counterForSystemCyclic(value);
+    case CSSCounterStyleDescriptors::System::Numeric:
+        return counterForSystemNumeric(value);
+    case CSSCounterStyleDescriptors::System::Alphabetic:
+        return counterForSystemAlphabetic(value);
+    case CSSCounterStyleDescriptors::System::Symbolic:
+        return counterForSystemSymbolic(value);
+    case CSSCounterStyleDescriptors::System::Additive:
+        return counterForSystemAdditive(value);
+    case CSSCounterStyleDescriptors::System::Fixed:
+        return counterForSystemFixed(value);
+    case CSSCounterStyleDescriptors::System::Extends:
+        // CounterStyle with extends system should have been promoted to another system at this point
+        ASSERT_NOT_REACHED();
+        break;
+    }
     return ""_s;
 }
+
+String CSSCounterStyle::fallbackText(int value)
+{
+    if (m_isFallingBack || !fallback().get()) {
+        m_isFallingBack = false;
+        return CSSCounterStyleRegistry::decimalCounter()->text(value);
+    }
+    m_isFallingBack = true;
+    auto fallbackText = fallback()->text(value);
+    m_isFallingBack = false;
+    return fallbackText;
```

Comment:
Is this `m_isFallingBack` field useful? Doesn't `fallback()->text()` eventually call `fallbackText()` again anyway?

Reply:
Yes, because we need to detect loops and fallback to decimal in that case